File: bf.zig
// Language: Zig
// Run with the Zig compiler like `zig run bf.zig`;
// compatible with v8.x and v9.x.

const std = @import("std");
const process = std.process;
const Allocator = std.mem.Allocator;
const print = std.debug.print;
const mem = std.mem;

fn interpret(stdin: anytype, stdout: anytype, code: []const u8) !void {
    var tape = [_]u8{0} ** 30_000;
    var pdata: u32 = 0;
    var pinst: u32 = 0;
    while (pinst < code.len) {
        var inst = code[pinst];
        switch (inst) {
            '>' => {
                if (pdata < tape.len - 1) pdata += 1;
            },
            '<' => {
                if (pdata > 0) pdata -= 1;
            },
            '+' => {
                tape[pdata] += 1;
            },
            '-' => {
                tape[pdata] -%= 1;
            },
            '.' => {
                try stdout.writeByte(tape[pdata]);
            },
            ',' => {
                _ = try stdout.write("Input>");
                tape[pdata] = stdin.readByte() catch tape[pdata];
            },
            '[' => {
                if (tape[pdata] == 0) {
                    var bracket_stack: u32 = 0;
                    while (pinst < code.len) {
                        pinst += 1;
                        if (code[pinst] == '[') {
                            bracket_stack += 1;
                        } else if (code[pinst] == ']' and bracket_stack > 0) {
                            bracket_stack -= 1;
                        } else if (code[pinst] == ']' and bracket_stack == 0) {
                            pinst += 1;
                            break;
                        }
                    }
                }
            },
            ']' => {
                if (tape[pdata] > 0) {
                    var bracket_stack: u32 = 0;
                    while (pinst > 0) {
                        pinst -= 1;
                        if (code[pinst] == ']') {
                            bracket_stack += 1;
                        } else if (code[pinst] == '[' and bracket_stack > 0) {
                            bracket_stack -= 1;
                        } else if (code[pinst] == '[' and bracket_stack == 0) {
                            break;
                        }
                    }
                }
            },
            else => {},
        }
        pinst += 1;
    }
}

const program = "++++++++[>++++[>++>+++>+++>+<<<<-]>+>+>->>+[<]<-]>>.>---.+++++++..+++.>>.<-.<.+++.------.--------.>>+.";

fn readLine(stdin: anytype, buffer: []u8) []u8 {
    return stdin.readUntilDelimiterOrEof(buffer, '\n') catch buffer orelse buffer;
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    var alloc = &gpa.allocator;
    const args = try process.argsAlloc(alloc);
    const stdin  = std.io.getStdIn().reader();
    const stdout = std.io.getStdOut().writer();
    if (args.len > 1) {
        try interpret(stdin, stdout, args[1]);
    } else {
        while (true) {
            try stdout.writeByte('>');
            var buffer: [1024]u8 = undefined;
            const data = readLine(stdin, buffer[0..]);
            if (data.len == 0) continue;
            if (mem.eql(u8, data, "quit")) {
                break;
            } else {
                try interpret(stdin, stdout, data);
            }
            try stdout.writeByte('\n');
        }
    }
}

File: bf.go
package main

import (
	"fmt"
	"os"
)

func interpret(code string) {
	tape := [3e5]byte{}
	pinst := 0
	pdata := 0
	for pinst < len(code) {
		switch code[pinst] {
		case '>':
			if pdata < len(tape)-1 {
				pdata += 1
			}
		case '<':
			if pdata > 0 {
				pdata -= 1
			}
		case '+':
			if tape[pdata] < 255 {
				tape[pdata] += 1
			}
		case '-':
			if tape[pdata] > 0 {
				tape[pdata] -= 1
			}
		case '.':
			fmt.Printf("%c", tape[pdata])
		case ',':
			fmt.Print("Input>")
			var input string
			n, err := fmt.Scanln(&input)
			if err != nil {
				fmt.Print(n, err)
			} else {
				tape[pdata] = input[0]
			}
		case '[':
			if tape[pdata] == 0 {
				var bracketStack uint32 = 0
				for {
					pinst += 1
					if code[pinst] == '[' {
						bracketStack += 1
					} else if code[pinst] == ']' && bracketStack > 0 {
						bracketStack -= 1
					} else if code[pinst] == ']' && bracketStack == 0 {
						pinst += 1
						break
					}
				}
			}
		case ']':
			if tape[pdata] != 0 {
				var bracketStack uint32 = 0
				for {
					pinst -= 1
					if code[pinst] == ']' {
						bracketStack += 1
					} else if code[pinst] == '[' && bracketStack > 0 {
						bracketStack -= 1
					} else if code[pinst] == '[' && bracketStack == 0 {
						break
					}
				}
			}
		}
		pinst += 1
	}
}

func main() {
	if len(os.Args) > 1 {
		interpret(os.Args[1])
	} else {
		for {
			fmt.Print(">")
			var input string
			n, err := fmt.Scanln(&input)
			if err != nil {
				fmt.Print(n, err)
				continue
			}
			interpret(input)
		}
	}
}
